### PR TITLE
Unique ID support in Core, React and Material

### LINKED ISSUE #691: Control IDs should be unique and escaped
IDs are not unique since the schema path is used for them currently. Also, if we use a variation of the schema path, we should make sure to escape it.

### PR DESCRIPTION
**Pull request by sdirix**
Fixes #691 

Content:
 * Adds unique ID support in ```core```
 * Uses unique ID support in ```react``` to provide IDs
 * Uses generated IDs in ```react``` to derive unique IDs for HTML elements in Material renderers

Open Tasks:
 * Implement tests for ALL material fields
 * Add unique ID support for arrays

### CODE REVIEW

File: packages/core/src/util/ids.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  let tries = 1;
+  while (!isUniqueId(proposedID, tries)) {
+    tries++;
+  }
+  const newID = makeId(proposedID, tries);
+  usedIDs.add(newID);
+  return newID;
+};
+
+export const removeId = id => usedIDs.delete(id);
```
Comment (edgarmueller) [inline packages/core/src/util/ids.ts]: Add type annotation to `id`

File: packages/react/src/JsonForms.tsx [hunk 2]
```diff
@@ -26,26 +26,57 @@ import * as React from 'react';
...
 import { UnknownRenderer } from './UnknownRenderer';
-import { JsonFormsProps, mapStateToDispatchRendererProps } from '@jsonforms/core';
+import {
+  createId,
+  JsonFormsProps,
+  mapStateToDispatchRendererProps,
+  removeId
+} from '@jsonforms/core';
+
+class JsonFormsDispatchRenderer extends React.Component<JsonFormsProps, { id: string }> {
```
Comment (edgarmueller) [inline packages/react/src/JsonForms.tsx]: Extract state to explicit type

File: packages/react/src/JsonForms.tsx [hunk 3]
```diff
@@ -26,26 +26,57 @@ import * as React from 'react';
...
+  }
 
-const JsonFormsDispatchRenderer =
-  ({ uischema, schema, path, renderers }: JsonFormsProps) => {
-  const renderer = _.maxBy(renderers, r => r.tester(uischema, schema));
-  if (renderer === undefined || renderer.tester(uischema, schema) === -1) {
-    return <UnknownRenderer type={'renderer'}/>;
-  } else {
-    const Render = renderer.renderer;
+  isManageOwnID(props) {
```
Comment (edgarmueller) [inline packages/react/src/JsonForms.tsx]: Re-use `isControl`

File: packages/react/test/renderers/JsonForms.test.tsx [hunk 4]
```diff
@@ -186,3 +210,61 @@ test('deregister an unregistered renderer should be a no-op', t => {
...
+        {children}
+      </div>
+    );
+  };
+
+  const store = initJsonFormsStore(
+    {
+      data: t.context.data,
+      schema: t.context.schema,
+      uischema: t.context.uischema2,
```
Comment (edgarmueller) [inline packages/react/test/renderers/JsonForms.test.tsx]: Make `uischema2` a local var as it doesn't seem to be used anywhere else.

**Comment by edgarmueller** [comment]
For the remaining open tasks, let's create corresponding issues.

File: packages/react/src/DispatchField.tsx [hunk 5]
```diff
@@ -27,31 +27,30 @@ import { connect } from 'react-redux';
...
-   * The available field renderers.
-   */
-
-const Dispatch = (dispatchFieldProps: DispatchFieldProps) => {
-  const uischema = dispatchFieldProps.uischema;
-  const schema = dispatchFieldProps.schema;
-  const field = _.maxBy(dispatchFieldProps.fields, r => r.tester(uischema, schema));
+class Dispatch extends React.Component<DispatchFieldProps> {
+  render() {
+    const { uischema, schema, path, fields, id } = this.props as DispatchFieldProps;
```
Comment (eneufeld) [inline packages/react/src/DispatchField.tsx]: I think the cast is not needed

**Comment by edgarmueller** [comment]
Closing in favor of #1050

**Pull request closed without merging**